x-pack/plugins/enterprise_search/public/applications/app_search/components/sample_engine_creation_cta/sample_engine_creation_cta_logic.ts

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { generatePath } from 'react-router-dom';
+
+import { kea, MakeLogicType } from 'kea';
+
+import { flashAPIErrors, setQueuedSuccessMessage } from '../../../shared/flash_messages';
+import { HttpLogic } from '../../../shared/http';
+import { KibanaLogic } from '../../../shared/kibana';
+import { ENGINE_PATH } from '../../routes';
+import { ENGINE_CREATION_SUCCESS_MESSAGE } from '../engine_creation/constants';
+
+interface SampleEngineCreationCtaActions {
+  createSampleEngine(): void;
+  onSampleEngineCreationSuccess(): void;
+  setIsLoading(isLoading: boolean): { isLoading: boolean };
+}
+
+interface SampleEngineCreationCtaValues {
+  isLoading: boolean;
+}
+
+export const SampleEngineCreationCtaLogic = kea<
+  MakeLogicType<SampleEngineCreationCtaValues, SampleEngineCreationCtaActions>
+>({
+  path: ['enterprise_search', 'app_search', 'sample_engine_cta_logic'],
+  actions: {
+    createSampleEngine: true,
+    onSampleEngineCreationSuccess: true,
+    setIsLoading: (isLoading) => ({ isLoading }),
```

Comment:
Actions like this are generally an [anti-pattern in Kea](https://kea.js.org/docs/guide/concepts#listeners).

It is so pervasive in our code that I generally ignore it in reviews. However, since this logic is small (and new?) I want to point it out. When we design logic files we should keep this mind.

The preferable way to do this is:
```
 actions: {
    createSampleEngine: true,
    onSampleEngineCreationSuccess: true,
    onSampleEngineCreationFailure: true,
  },
  reducers: {
    isLoading: [
    false,
    {
      createSampleEngine: () => true,
      onSampleEngineCreationSuccess: () => false,
      onSampleEngineCreationFailure: () => false,
    },
  ],
  },
    ....
      try {
        await http.post('/api/app_search/onboarding_complete', {
          body,
        });
        actions.onSampleEngineCreationSuccess();
      } catch (e) {
        actions.onSampleEngineCreationFailure();
        flashAPIErrors(e);
      }

```

Reply:
I actually like this a lot more!

Reply:
Sweet. And just FYI, after re-reading this I don't know if I was totally clear about *what* the anti-pattern is. From that link:

> If you read the reducers section above, you'll remember that it's an anti-pattern to only have setThis and setThat actions that only update this or that.

> A helpful mental model to understand actions is the one of events in computer programming.

This is something I've mentioned on a couple of PRs previously, but would be new to you:

https://github.com/elastic/kibana/pull/83353#discussion_r524412310

https://github.com/elastic/kibana/pull/86822#pullrequestreview-557914749

Some lengthy discussion in there!